### DISCUSSION TITLE
How do I set an attribute on a span where the value is a map or struct?

### DISCUSSION THREAD

**Discussion by theFong** [open-telemetry/opentelemetry-go#4878] [Q&A]
I've used structured loggers like zap and logrus where I can set maps or structs. How can I do this with opentelemetry traces? I feel like it is possible since I am using datadog and https://github.com/connectrpc/otelconnect-go which results in traces with nested attributes that look like this in datadog. (Apologies if this is too datadog specific)
<img width="151" alt="image" src="https://github.com/open-telemetry/opentelemetry-go/assets/14320477/2c8e0c24-cce3-41cd-a728-23ef2a8d9e05">

However, I can't seem to find anything in this package that looks like `attribute.Map` or `attribute.Any`.
What strategies are people using?

**Comment by dmathieu** [comment]
There is an attribute.StringSlice for arrays of strings.
https://pkg.go.dev/go.opentelemetry.io/otel/attribute#StringSlice

IntSlice exists too.

Maps are a bit different, as otel doesn't explicitly say how nested hash keys should be represented (they will have to end up being flattened). So you will have to setup your own helper method which will loop through each entry and set the proper key/value.

**Reply by theFong** [reply]
Thanks for the response! So if a had a map that looked like:
```
map[string]any{
	"first": map[string]string {
		"second": "value",
	},
}
```
what would my helper func turn this into?

**Reply by dmathieu** [reply]
You could have one attribute with the key "first.second", and the value "value".

**Comment by pellared** [comment]
AFAIK most people do marshal JSON and pass it as a string attribute.

Also a heads up, that OTel Go Logs (which is in design phase) will allow complex, structured attributes.

**Reply by theFong** [reply]
Thanks! We're currently json mashalling and storing as string but am noticing that in Datadog it's showing up as a string yet other attributes are properly identified as nested. The library that is properly adding the nested attributes though is an otel implementation so it's not something that's datadog specific. This is important because it makes it really easy to filter for traces etc in UI.

**Reply by pellared** [reply]
I guess that as a Datadog user/customer you can ask them for help.
